File: truth_set/python/structure/organisation/complex-code/e-commerce.py
from enum import Enum
from typing import Dict, List, Optional, Any
from datetime import datetime

class LoyaltyStatus(Enum):
    NONE = "NONE"
    SILVER = "SILVER" 
    GOLD = "GOLD"
    PLATINUM = "PLATINUM"

class ShippingMethod(Enum):
    STANDARD = "STANDARD"
    EXPRESS = "EXPRESS"
    OVERNIGHT = "OVERNIGHT"

def process_order(order: Dict) -> Optional[Dict]:
    final_order = {}
    
    if not (order and order.get('customer') and order['customer'].get('id') and 
            order.get('items') and len(order['items']) > 0):
        return None
        
    final_order['customerId'] = order['customer']['id']
    final_order['customerName'] = order['customer'].get('name', 'Valued Customer')
    
    if order['customer'].get('loyaltyStatus') and order['customer']['loyaltyStatus'] != LoyaltyStatus.NONE.value:
        final_order['loyalty'] = f"Status: {order['customer']['loyaltyStatus']}"
        
    final_order['items'] = {}
    
    for i, item in enumerate(order['items']):
        if not (item.get('productId') and item.get('quantity') and item['quantity'] > 0):
            continue
            
        item_total = 0
        product_price = get_product_price(item['productId'])
        
        if not product_price:
            final_order[f'item_{i}'] = 'Price Not Available'
            continue
            
        item_total += product_price * item['quantity']
        
        if item.get('options') and len(item['options']) > 0:
            for option in item['options']:
                if not (option.get('optionId') and option.get('value')):
                    continue
                    
                option_price = get_option_price(option['optionId'], option['value'])
                if option_price:
                    item_total += option_price
                    
        if item.get('discounts') and len(item['discounts']) > 0:
            for discount in item['discounts']:
                if not (discount.get('discountId') and discount.get('amount') and discount['amount'] > 0):
                    continue
                item_total -= discount['amount']
                
        final_order['items'][f'item_{i}'] = {
            'productId': item['productId'],
            'quantity': item['quantity'],
            'total': max(item_total, 0)
        }
        
        if order.get('coupons') and len(order['coupons']) > 0:
            for coupon in order['coupons']:
                if not (coupon.get('code') and coupon.get('valid') and 
                        datetime.strptime(coupon['expiryDate'], '%Y-%m-%d') > datetime.now()):
                    continue
                final_order['items'][f'item_{i}']['coupon'] = coupon['code']
                
    if order.get('shipping') and order['shipping'].get('method'):
        if not (order['shipping'].get('cost') and order['shipping']['cost'] > 0):
            final_order['shipping'] = f"Method: {order['shipping']['method']}, Cost: Calculating"
        else:
            final_order['shipping'] = f"Method: {order['shipping']['method']}, Cost: ${order['shipping']['cost']}"
    else:
        final_order['shipping'] = 'Standard Shipping'
        
    if not final_order['items']:
        return None
        
    return final_order

def get_product_price(product_id: str) -> Optional[float]:
    price_list = {
        'P001': 29.99,
        'P002': 49.99,
        'P003': 19.99
    }
    return price_list.get(product_id)

def get_option_price(option_id: str, value: str) -> Optional[float]:
    option_price_list = {
        'O001': {'Red': 5.0, 'Blue': 5.0},
        'O002': {'Large': 10.0, 'Medium': 0.0}
    }
    return option_price_list.get(option_id, {}).get(value)

sample_order = {
    'customer': {
        'id': 'CUST12345',
        'name': 'Jane Smith',
        'loyaltyStatus': 'GOLD'
    },
    'items': [
        {
            'productId': 'P001',
            'quantity': 3,
            'options': [
                {'optionId': 'O001', 'value': 'Red'},
                {'optionId': 'O002', 'value': 'Large'}
            ],
            'discounts': [
                {'discountId': 'D001', 'amount': 10},
                {'discountId': 'D002', 'amount': 5}
            ]
        },
        {
            'productId': 'P004',
            'quantity': 2,
            'options': [
                {'optionId': 'O003', 'value': 'Green'}
            ],
            'discounts': [
                {'discountId': 'D003', 'amount': -5}
            ]
        },
        {
            'quantity': 1,
            'options': [
                {'optionId': 'O001', 'value': 'Blue'}
            ],
            'discounts': [
                {'discountId': 'D004', 'amount': 3}
            ]
        }
    ],
    'shipping': {
        'address': '456 Elm Street, Othertown, USA',
        'method': 'EXPRESS',
        'cost': 20.0
    },
    'coupons': [
        {'code': 'WELCOME10', 'valid': True, 'expiryDate': '2025-01-01'},
        {'code': 'BLACKFRIDAY', 'valid': False, 'expiryDate': '2023-11-30'}
    ]
}

processed_order = process_order(sample_order)
print('Processed Order:', processed_order)

File: truth_set/javascript/structure/organisation/complex-code/e-commerce.js
var LoyaltyStatus;
(function (LoyaltyStatus) {
    LoyaltyStatus["NONE"] = "NONE";
    LoyaltyStatus["SILVER"] = "SILVER";
    LoyaltyStatus["GOLD"] = "GOLD";
    LoyaltyStatus["PLATINUM"] = "PLATINUM";
})(LoyaltyStatus || (LoyaltyStatus = {}));
var ShippingMethod;
(function (ShippingMethod) {
    ShippingMethod["STANDARD"] = "STANDARD";
    ShippingMethod["EXPRESS"] = "EXPRESS";
    ShippingMethod["OVERNIGHT"] = "OVERNIGHT";
})(ShippingMethod || (ShippingMethod = {}));
function processOrder(order) {
    var finalOrder = {};
    if (!(order && order.customer && order.customer.id && order.items && order.items.length > 0)) {
        return null;
    }
    finalOrder['customerId'] = order.customer.id;
    finalOrder['customerName'] = order.customer.name || 'Valued Customer';
    if (order.customer.loyaltyStatus && order.customer.loyaltyStatus !== LoyaltyStatus.NONE) {
        finalOrder['loyalty'] = "Status: ".concat(order.customer.loyaltyStatus);
    }
    finalOrder['items'] = {};
    for (var i = 0; i < order.items.length; i++) {
        var item = order.items[i];
        if (!(item.productId && item.quantity && item.quantity > 0)) {
            continue;
        }
        var itemTotal = 0;
        var productPrice = getProductPrice(item.productId);
        if (!(productPrice)) {
            finalOrder["item_".concat(i)] = 'Price Not Available';
            continue;
        }
        itemTotal += productPrice * item.quantity;
        if (item.options && item.options.length > 0) {
            for (var j = 0; j < item.options.length; j++) {
                var option = item.options[j];
                if (!(option.optionId && option.value)) {
                    continue;
                }
                var optionPrice = getOptionPrice(option.optionId, option.value);
                if (optionPrice) {
                    itemTotal += optionPrice;
                }
            }
        }
        if (item.discounts && item.discounts.length > 0) {
            for (var k = 0; k < item.discounts.length; k++) {
                var discount_1 = item.discounts[k];
                if (!(discount_1.discountId && discount_1.amount && discount_1.amount > 0)) {
                    continue;
                }
                itemTotal -= discount_1.amount;
            }
        }
        finalOrder['items']["item_".concat(i)] = {
            productId: item.productId,
            quantity: item.quantity,
            total: itemTotal > 0 ? itemTotal : 0,
        };
        if (order.coupons && order.coupons.length > 0) {
            for (var c = 0; c < order.coupons.length; c++) {
                var coupon = order.coupons[c];
                if (!(coupon.code && coupon.valid && new Date(coupon.expiryDate) > new Date())) {
                    continue;
                }
                finalOrder["items"]["item_".concat(i)]['coupon'] = coupon.code;
            }
        }
    }
    if (order.shipping && order.shipping.method) {
        if (!(order.shipping.cost && order.shipping.cost > 0)) {
            finalOrder['shipping'] = "Method: ".concat(order.shipping.method, ", Cost: Calculating");
        }
        else {
            finalOrder['shipping'] = "Method: ".concat(order.shipping.method, ", Cost: $").concat(order.shipping.cost);
        }
    }
    else {
        finalOrder['shipping'] = 'Standard Shipping';
    }
    if (!Object.keys(finalOrder['items']).length) {
        return undefined;
    }
    return finalOrder;
}
function getProductPrice(productId) {
    var priceList = {
        'P001': 29.99,
        'P002': 49.99,
        'P003': 19.99,
    };
    return priceList[productId] || null;
}
function getOptionPrice(optionId, value) {
    var _a;
    var optionPriceList = {
        'O001': { 'Red': 5.0, 'Blue': 5.0 },
        'O002': { 'Large': 10.0, 'Medium': 0.0 },
    };
    return ((_a = optionPriceList[optionId]) === null || _a === void 0 ? void 0 : _a[value]) || null;
}
var sampleOrder = {
    customer: {
        id: 'CUST12345',
        name: 'Jane Smith',
        loyaltyStatus: 'GOLD',
    },
    items: [
        {
            productId: 'P001',
            quantity: 3,
            options: [
                { optionId: 'O001', value: 'Red' },
                { optionId: 'O002', value: 'Large' },
            ],
            discounts: [
                { discountId: 'D001', amount: 10 },
                { discountId: 'D002', amount: 5 },
            ],
        },
        {
            productId: 'P004',
            quantity: 2,
            options: [
                { optionId: 'O003', value: 'Green' },
            ],
            discounts: [
                { discountId: 'D003', amount: -5 },
            ],
        },
        {
            quantity: 1,
            options: [
                { optionId: 'O001', value: 'Blue' },
            ],
            discounts: [
                { discountId: 'D004', amount: 3 },
            ],
        },
    ],
    shipping: {
        address: '456 Elm Street, Othertown, USA',
        method: 'EXPRESS',
        cost: 20.0,
    },
    coupons: [
        { code: 'WELCOME10', valid: true, expiryDate: '2025-01-01' },
        { code: 'BLACKFRIDAY', valid: false, expiryDate: '2023-11-30' },
    ],
};
var processedOrder = processOrder(sampleOrder);
console.log('Processed Order:', processedOrder);